File: samlang-checker/src/main/kotlin/samlang/checker/TypeResolver.kt
package samlang.checker

import samlang.ast.common.Type
import samlang.ast.common.Type.FunctionType
import samlang.ast.common.Type.IdentifierType
import samlang.ast.common.Type.PrimitiveType
import samlang.ast.common.Type.TupleType
import samlang.ast.common.Type.UndecidedType
import samlang.ast.common.TypeVisitor

private typealias ResolveTypeFunction = (undecidedType: UndecidedType) -> Type

internal fun resolveType(type: Type, function: ResolveTypeFunction): Type =
    type.accept(visitor = TypeResolverVisitor, context = function)

private object TypeResolverVisitor :
    TypeVisitor<ResolveTypeFunction, Type> {

    override fun visit(type: PrimitiveType, context: ResolveTypeFunction): Type = type

    override fun visit(type: IdentifierType, context: ResolveTypeFunction): Type =
        type.copy(typeArguments = type.typeArguments.map { it.accept(visitor = this, context = context) })

    override fun visit(type: TupleType, context: ResolveTypeFunction): Type =
        type.copy(mappings = type.mappings.map { it.accept(visitor = this, context = context) })

    override fun visit(type: FunctionType, context: ResolveTypeFunction): Type =
        type.copy(
            argumentTypes = type.argumentTypes.map { it.accept(visitor = this, context = context) },
            returnType = type.returnType.accept(visitor = this, context = context)
        )

    override fun visit(type: UndecidedType, context: ResolveTypeFunction): Type = context(type)
}

File: typescript-rewrite/samlang-core/src/checker/type-resolver.ts
import { Type, UndecidedType, identifierType, tupleType, functionType } from '../ast/common/types';

const resolveType = (
  type: Type,
  undecidedTypeResolver: (undecidedType: UndecidedType) => Type
): Type => {
  switch (type.type) {
    case 'PrimitiveType':
      return type;
    case 'IdentifierType':
      return identifierType(
        type.identifier,
        type.typeArguments.map((it) => resolveType(it, undecidedTypeResolver))
      );
    case 'TupleType':
      return tupleType(type.mappings.map((it) => resolveType(it, undecidedTypeResolver)));
    case 'FunctionType':
      return functionType(
        type.argumentTypes.map((it) => resolveType(it, undecidedTypeResolver)),
        resolveType(type.returnType, undecidedTypeResolver)
      );
    case 'UndecidedType':
      return undecidedTypeResolver(type);
  }
};

export default resolveType;
